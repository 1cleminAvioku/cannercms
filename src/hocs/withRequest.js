// @flow
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {generateAction} from 'provider';
import isArray from 'lodash/isArray';
import type {RelationDef} from './relationFactory';
import {create} from './relationFactory';

type Props = {
  id: string,
  isEntity: boolean,
  name: string,
  title: string, // only used in withTitleAndDesription hoc
  description: string, // only used in withTitleAndDesription hoc
  layout?: 'horizontal' | 'vertical', // only used in withTitleAndDesription hoc
  [string]: any,
  rootValue: any,
  relation?: RelationDef
};

type changeQueue = Array<{id: string | {firstId: string, secondId: string}, type: any, value: any}>;

// $FlowFixMe
export default function withRequest(Com) {
  // this hoc will update data;
  // it's the onChange end
  return class ComponentWithRequest extends Component<Props> {
    static contextTypes = {
      request: PropTypes.func,
    };

    onChange = (id: string | {firstId: string, secondId: string} | changeQueue, type: any, delta: any) => {
      if (isArray(id)) {
        const changeQueue = id;
        // $FlowFixMe
        changeQueue.forEach((args) => {
          const {id, type, value} = args;
          this.onChange(id, type, value);
        });
        return;
      }
      const {request} = this.context;
      const {rootValue, relation, value} = this.props;
      // generate action
      const action = createAction(relation, id, type, delta, rootValue, value);
      if (!action) {
        return;
      }

      // quick and temp solution
      if (relation && relation.relationship === 'oneToMany.foreignKey') {
        request(action).then(() => this.props.fetchRelation());
      } else {
        request(action);
      }
    }

    render() {
      // default onChange function is pass by this hoc not parent
      // but in some situation maybe parent plugin want control the data like popup modal.
      const {onChange} = this.props;
      return <Com {...this.props} onChange={onChange || this.onChange} />;
    }
  };
}

function createAction(relation, id, type, delta, rootValue, value) {
  const defaultReturn = () => generateAction(id, type, delta, rootValue);
  if (!relation) {
    return defaultReturn();
  }

  const relationInstance = create(relation);
  if (relation.relationship === 'oneToMany.foreignKey' && type === 'delete') {
    // give the correct rootValue
    rootValue = value;
  }

  if (relationInstance.transformValue) {
    return generateAction(id, type, relationInstance.transformValue(delta), rootValue);
  }

  if (relationInstance.createAction) {
    return relationInstance.createAction(id, type, delta, rootValue);
  }

  // default return
  return defaultReturn();
}
