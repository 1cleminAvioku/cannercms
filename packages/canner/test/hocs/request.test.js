import * as React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import withRequest, {createAction, loop} from '../../src/hocs/request';
import RefId from 'canner-ref-id';

Enzyme.configure({ adapter: new Adapter() });

describe('with request', () => {
  let WrapperComponent, props, MockComponent, mockRequest;

  beforeEach(() => {
    MockComponent = function MockComponent() {
      return (<div>Component</div>);
    }
    mockRequest = jest.fn().mockImplementation(() => Promise.resolve());
    const rootValue = {
      posts: [{
        id: 'id1',
        title: 'post1'
      }]
    };
    props = {
      refId: new RefId('posts'),
      value: rootValue,
      rootValue,
      relation: undefined,
      fetchRelation: () => {
      },
      items: {
        title: {
          type: 'string'
        }
      },
      schema: {
        posts: {
          type: 'array',
          keyName: 'posts',
          items: {
            type: 'object',
            items: {
              title: {
                keyName: 'title',
                type: 'string'
              }
            }
          }
        }
      },
      request: mockRequest
    }
    WrapperComponent = withRequest(MockComponent);
  });

  it('should render', () => {
    const wrapper = shallow(<WrapperComponent {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('onChang function', () => {
    it('should call props.request with arguments action', () => {
      const wrapper = shallow(<WrapperComponent {...props} />);
      const type = 'create';
      const delta = {
        id: 'id2',
        title: 'post2'
      };
      wrapper.instance().onChange(props.refId, type, delta).then(() => {
        expect(mockRequest.mock.calls[0][0]).toBeTruthy();
      });
    });

    it('should accept changeQueue', () => {
      const wrapper = shallow(<WrapperComponent {...props} />);
      const refId = props.refId;
      const type = 'create';
      const delta = i => ({
        title: `post${i}`
      });
      const changeQueue = [2,3,4,5].map(i => ({refId, type, value: delta(i)}));
      return wrapper.instance().onChange(changeQueue)
        .then(() => {
          expect(mockRequest.mock.calls.length).toBe(4);
        });
    });
  });
});


describe('createAction', () => {
  describe('create onChange', () => {
    it('create with delta should merge with emptyData', () => {
      const action = createAction({
        id: 'posts',
        type: 'create',
        value: {
          id: 'id2',
          title: 'post2'
        },
        rootValue: {posts: []},
        items: {
          title: {
            type: 'string'
          },
          field: {
            type: 'number'
          }
        }
      })
      expect(action.type).toBe('CREATE_ARRAY');
      expect(action.payload.value).toMatchObject({
        title: 'post2',
        field: 0
      });
    });

    it('create with delta should not merge with emptyData', () => {
      const action = createAction({
        id: 'posts',
        type: 'create',
        value: {
          id: 'id2',
          title: 'post2'
        },
        config: true, // this will cause unmerged
        rootValue: {posts: []},
        items: {
          title: {
            type: 'string'
          },
          field: {
            type: 'number'
          }
        }
      })
      expect(action.type).toBe('CREATE_ARRAY');
      expect(action.payload.value).toMatchObject({
        title: 'post2'
      });
    });

    it('create without delta should create emptyData', () => {
      const action = createAction({
        id: 'posts',
        type: 'create',
        config: false, // this will cause unmerged
        rootValue: {posts: []},
        items: {
          title: {
            type: 'string'
          },
          field: {
            type: 'number'
          }
        }
      })
      expect(action.type).toBe('CREATE_ARRAY');
      expect(action.payload.value).toMatchObject({
        title: '',
        field: 0
      });
    });
  });
});

describe('loop schema', () => {
  const schema = {
    info: {
      type: 'object',
      keyName: 'info',
      items: {
        name: {
          type: 'string',
          keyName: 'name',
        }
      }
    },
    posts: {
      type: 'array',
      keyName: 'posts',
      items: {
        type: 'object',
        items: {
          name: {
            keyName: 'name',
            type: 'string'
          }
        }
      }
    }
  }

  it('should find info', () => {
    const paths = ['info'];
    expect(loop(schema, paths, '')).toMatchObject({
      type: 'object',
      keyName: 'info',
      pattern: 'object'
    });
  });

  it('should find posts', () => {
    const paths = ['posts'];
    expect(loop(schema, paths, '')).toMatchObject({
      type: 'array',
      keyName: 'posts',
      pattern: 'array'
    });
  });

  it('should find info/name', () => {
    const paths = ['info', 'name'];
    expect(loop(schema, paths, '')).toMatchObject({
      type: 'string',
      keyName: 'name',
      pattern: 'object/string'
    });
  });

  it('should find posts/0/name', () => {
    const paths = ['posts', '0', 'name'];
    expect(loop(schema, paths, '')).toMatchObject({
      type: 'string',
      keyName: 'name',
      pattern: 'array/string'
    });
  });

})
