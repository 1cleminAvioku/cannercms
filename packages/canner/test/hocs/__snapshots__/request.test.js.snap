// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with request should render 1`] = `
<MockComponent
  fetchRelation={[Function]}
  items={
    Object {
      "title": Object {
        "tyep": "string",
      },
    }
  }
  onChange={[Function]}
  refId={
    CannerRefId {
      "_internalArr": Array [
        "posts",
      ],
      "child": [Function],
      "getPathArr": [Function],
      "remove": [Function],
      "toString": [Function],
    }
  }
  request={[MockFunction]}
  rootValue={
    Object {
      "posts": Array [
        Object {
          "id": "id1",
          "title": "post1",
        },
      ],
    }
  }
  value={
    Object {
      "posts": Array [
        Object {
          "id": "id1",
          "title": "post1",
        },
      ],
    }
  }
/>
`;
